op-node/node/node.go

```diff
@@ -382,148 +478,160 @@ func (n *OpNode) initL1BeaconAPI(ctx context.Context, cfg *config.Config) error
+	altDA := altda.NewAltDA(node.log, cfg.AltDA, rpCfg, node.metrics.AltDAMetrics)
+	var safeDB closableSafeDB
 	if cfg.SafeDBPath != "" {
-		n.log.Info("Safe head database enabled", "path", cfg.SafeDBPath)
-		safeDB, err := safedb.NewSafeDB(n.log, cfg.SafeDBPath)
+		node.log.Info("Safe head database enabled", "path", cfg.SafeDBPath)
+		safeDB, err = safedb.NewSafeDB(node.log, cfg.SafeDBPath)
 		if err != nil {
-			return fmt.Errorf("failed to create safe head database at %v: %w", cfg.SafeDBPath, err)
+			return nil, nil, nil, nil, fmt.Errorf("failed to create safe head database at %v: %w", cfg.SafeDBPath, err)
 		}
-		n.safeDB = safeDB
 	} else {
-		n.safeDB = safedb.Disabled
+		safeDB = safedb.Disabled
 	}
 
 	if cfg.Rollup.ChainOpConfig == nil {
-		return fmt.Errorf("cfg.Rollup.ChainOpConfig is nil. Please see https://github.com/ethereum-optimism/optimism/releases/tag/op-node/v1.11.0: %w", err)
+		return nil, nil, nil, nil, fmt.Errorf("cfg.Rollup.ChainOpConfig is nil. Please see https://github.com/ethereum-optimism/optimism/releases/tag/op-node/v1.11.0: %w", err)
 	}
 
-	n.l2Driver = driver.NewDriver(n.eventSys, n.eventDrain, &cfg.Driver, &cfg.Rollup, cfg.L1ChainConfig, cfg.DependencySet, n.l2Source, n.l1Source,
-		n.beacon, n, n, n.log, n.metrics, cfg.ConfigPersistence, n.safeDB, &cfg.Sync, sequencerConductor, altDA, indexingMode)
+	l2Driver := driver.NewDriver(node.eventSys, node.eventDrain, &cfg.Driver, &cfg.Rollup, cfg.L1ChainConfig, cfg.DependencySet, l2Source, node.l1Source,
+		node.beacon, node, node, node.log, node.metrics, cfg.ConfigPersistence, safeDB, &cfg.Sync, sequencerConductor, altDA, indexingMode)
 
 	// Wire up IndexingMode to engine controller for direct procedure call
-	if n.interopSys != nil {
-		if indexingMode, ok := n.interopSys.(*indexing.IndexingMode); ok {
-			indexingMode.SetEngineController(n.l2Driver.SyncDeriver.Engine)
+	if sys != nil {
+		if indexingMode, ok := sys.(*indexing.IndexingMode); ok {
+			indexingMode.SetEngineController(l2Driver.SyncDeriver.Engine)
 		}
 	}
 
-	return nil
+	return l2Source, sys, l2Driver, safeDB, nil
 }
 
-func (n *OpNode) initRPCServer(cfg *config.Config) error {
+func initRPCServer(cfg *config.Config, node *OpNode) (*oprpc.Server, error) {
 	server := newRPCServer(&cfg.RPC, &cfg.Rollup, cfg.DependencySet,
-		n.l2Source.L2Client, n.l2Driver, n.safeDB,
-		n.log, n.metrics, n.appVersion)
-	if p2pNode := n.getP2PNodeIfEnabled(); p2pNode != nil {
-		server.AddAPI(rpc.API{
+		node.l2Source.L2Client, node.l2Driver, node.safeDB,
+		node.log, node.metrics, node.appVersion)
+	if err := registerAPIs(cfg, node, server.Handler); err != nil {
+		// panic here is to match the behavior of oprcp.Server.AddAPI,
+		// which wraps the Handler and panics if the API can't be added.
+		panic(fmt.Errorf("invalid API: %w", err))
+	}
+	node.log.Info("Starting JSON-RPC server")
+	if err := server.Start(); err != nil {
+		return nil, fmt.Errorf("unable to start RPC server: %w", err)
+	}
```

Comment:
Ah, I see this ordering would be hard to preserve with my earlier suggestion about just returning the handler.  It's kind of weird that we start the server here instead of in the `node.Start()` method but probably out of scope to try and change that in this PR.